File: docs/sdk-extras/cordapp/workflows/src/main/java/com/r3/conclave/cordapp/client/EnclaveClientHelper.java
package com.r3.conclave.cordapp.client;

import co.paralleluniverse.fibers.Suspendable;
import com.r3.conclave.client.EnclaveConstraint;
import com.r3.conclave.client.InvalidEnclaveException;
import com.r3.conclave.common.EnclaveInstanceInfo;
import com.r3.conclave.mail.EnclaveMail;
import com.r3.conclave.mail.PostOffice;
import net.corda.core.flows.FlowException;
import net.corda.core.flows.FlowSession;
import org.jetbrains.annotations.NotNull;
import java.io.IOException;

/**
 * A helper class that wraps some boilerplate for communicating with enclaves.
 */
public class EnclaveClientHelper {
    private final FlowSession session;
    private final String constraint;
    private PostOffice postOffice;

    public EnclaveClientHelper(FlowSession session, String constraint) {
        this.session = session;
        this.constraint = constraint;
    }

    @Suspendable
    public EnclaveClientHelper start() throws FlowException {
        if (postOffice != null)
            throw new IllegalStateException("An EnclaveClientHelper may not be started more than once.");

        // Read the enclave attestation from the peer.
        // In future, deserialization will be handled more automatically.
        EnclaveInstanceInfo attestation = session.receive(byte[].class).unwrap(EnclaveInstanceInfo::deserialize);

        // The key hash below (the hex string after 'S') is the public key version of sample_private_key.pem
        // In a real app you should remove the SEC:INSECURE part, of course.
        try {
            EnclaveConstraint.parse(constraint).check(attestation);
        } catch (InvalidEnclaveException e) {
            throw new FlowException(e);
        }

        // This will create a post office with a random sender key and use the "default" topic.
        postOffice = attestation.createPostOffice();
        return this;
    }

    @Suspendable
    public void sendToEnclave(byte[] messageBytes) {
        if (postOffice == null)
            throw new IllegalStateException("You must call start() first.");
        // Create the encrypted message and send it. We'll use a temporary key for now, so the message is
        // effectively unauthenticated so the enclave won't know who sent it. Future samples will show how to
        // integrate with the Corda identity infrastructure.
        //
        session.send(postOffice.encryptMail(messageBytes));
    }

    @Suspendable
    @NotNull
    public byte[] receiveFromEnclave() throws FlowException {
        if (postOffice == null)
            throw new IllegalStateException("You must call start() first.");
            EnclaveMail reply = session.receive(byte[].class).unwrap((mail) -> {
                try {
                    return postOffice.decryptMail(mail);
                } catch (IOException e) {
                    throw new FlowException(e);
                }
            });
            return reply.getBodyAsBytes();
    }
}

File: docs/sdk-extras/cordapp/workflows/src/main/kotlin/com/r3/conclave/cordapp/sample/client/EnclaveClientHelper.kt
package com.r3.conclave.cordapp.sample.client

import co.paralleluniverse.fibers.Suspendable
import com.r3.conclave.common.EnclaveConstraint
import com.r3.conclave.common.InvalidEnclaveException
import com.r3.conclave.common.EnclaveInstanceInfo
import com.r3.conclave.cordapp.host.EnclaveHostService
import net.corda.core.flows.FlowException
import net.corda.core.flows.FlowLogic
import net.corda.core.flows.FlowSession
import net.corda.core.identity.Party
import net.corda.core.utilities.unwrap

/**
 * A factory class that wraps some boilerplate for setting up secure initiator and receiver flow helpers
 */
object EnclaveClientHelper {
    /**
     * Creates a [EnclaveFlowInitiator] helper class instance that wraps some boilerplate for senders to initiate
     * secure flows and communicate with a given party.
     *
     * @param flow takes a reference to the flow being executed
     * @param receiver the receiving party in the network
     * @param constraint the Enclave constraints to be used to validate the other party's Enclave instance
     * @param anonymous whether the flow initiator should send its identity to the enclave or not
     * @return the instance of the flow initiator helper class
     * @throws FlowException
     */
    @Suspendable
    @Throws(FlowException::class)
    @JvmStatic
    @JvmOverloads
    fun initiateFlow(flow: FlowLogic<*>, receiver: Party, constraint: String,
                     anonymous: Boolean = false): EnclaveFlowInitiator {
        val session = flow.initiateFlow(receiver)

        // Read the enclave attestation from the peer.
        val attestation = session.receive(ByteArray::class.java).unwrap { from: ByteArray ->
            EnclaveInstanceInfo.deserialize(from)
        }

        // The key hash below (the hex string after 'S') is the public key version of sample_private_key.pem
        // In a real app you should remove the SEC:INSECURE part, of course.
        try {
            EnclaveConstraint.parse(constraint).check(attestation)
        } catch (e: InvalidEnclaveException) {
            throw FlowException(e)
        }
        val instance = EnclaveFlowInitiator(flow, session, attestation)
        instance.sendIdentityToEnclave(anonymous)

        return instance
    }

    /**
     * Creates a [EnclaveFlowResponder] helper class instance that wraps some boilerplate for receivers to initiate
     * secure flows and respond to a given party.
     * @param flow takes a reference to the flow being executed
     * @param counterPartySession the sender party
     * @param serviceType a class that specializes the [EnclaveHostService] type to define the enclave to be loaded
     * @return the instance of the responder flow responder helper class
     */
    @Suspendable
    @Throws(FlowException::class)
    @JvmStatic
    fun <T : EnclaveHostService> initiateResponderFlow(flow: FlowLogic<*>, counterPartySession: FlowSession,
                                                       serviceType: Class<T>): EnclaveFlowResponder {
        // Start an instance of the enclave hosting service
        val host = flow.serviceHub.cordaService(serviceType)
        // Send the other party the enclave identity (remote attestation) for verification.
        counterPartySession.send(host.attestationBytes)
        val instance = EnclaveFlowResponder(flow, counterPartySession, host)
        // Relay the initial identity message to the enclave and relay the response back
        instance.relayMessageToFromEnclave()
        return instance
    }
}
